File: 3_factorialize_num/factorialize.pl
#!/usr/bin/perl

use strict;
use warnings;

sub factorialize {
    my ($num) = @_;
    if ($num == 0) {
	return 1;
    } else {
	return $num * factorialize($num - 1);
    }
}

factorialize(0) == 1 || die "Bad return value for 1\n";
factorialize(5) == 120 || die "Bad return value for 5\n";
factorialize(7) == 5040 || die "Bad return value for 7\n";
factorialize(9) == 362880 || die "Bad return value for 9\n";

print "all tests passed\n";

File: 3_factorialize_num/factorialize.kt
fun main(args : Array<String>) {
    println(factorialize(7))
}

private fun factorialize(n: Int): Int {
    return when  {
        n == 0 -> 1
        n <= 1 -> n
        else -> n * factorialize(n - 1)
    }
}